doc/learn/syntax-trees-with-typescript.md

```diff
@@ -0,0 +1,523 @@
+Each of these different types extend `Node` (sometimes indirectly through `Literal` or `Parent`) and set `type` to a [string literal](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types) which uniquely identifies a kind of content (in TypeScript parlance a [discriminated union](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions)).
+
+A `Node` also can optionally include `Data`.
+`Data` is a dictionary/object which can store extra information and metadata which is not standard to a given node `type`.
+
+When a syntax tree is parsed from a file, it may include `Position`, which is information about where this `Node` is found in the source file.
+
+```ts
+/**
+ * Syntactic units in unist syntax trees are called nodes.
+ */
+interface Node {
+    /**
+     * The variant of a node.
+     */
+    type: string;
+
+    /**
+     * Information from the ecosystem.
+     */
+    data?: Data | undefined;
+
+    /**
+     * Location of a node in a source document.
+     * Must not be present if a node is generated.
+     */
+    position?: Position | undefined;
+}
+
+/**
+ * Information associated by the ecosystem with the node.
+ * Space is guaranteed to never be specified by unist or specifications
+ * implementing unist.
+ */
+export interface Data {
+    [key: string]: unknown;
+}
+
+/**
+ * Location of a node in a source file.
+ */
+export interface Position {
+    /**
+     * Place of the first character of the parsed source region.
+     */
+    start: Point;
+
+    /**
+     * Place of the first character after the parsed source region.
+     */
+    end: Point;
+
+    /**
+     * Start column at each index (plus start line) in the source region,
+     * for elements that span multiple lines.
+     */
+    indent?: number[] | undefined;
+}
+```
+
+#### Literal
+
+`Literal` extends `Node` and adds a `value` property.
+For example a Markdown `text` `Node` extends `Literal` and sets `value` to be a `string`.
+
+```ts
+/**
+ * Nodes containing a value.
+ */
+export interface Literal extends Node {
+    value: unknown;
+}
+```
+
+#### Parent
+
+
+```ts
+/**
+ * Nodes containing other nodes.
+ */
+export interface Parent extends Node {
+    /**
+     * List representing the children of a node.
+     */
+    children: Node[];
+}
+```
+
+#### Pulling unist into a project
+
+```bash
+npm install --save-dev @types/unist
+```
+
+then import the types into a TypeScript file with:
+
+```ts
+import type {Node, Literal, Parent} from 'unist'
+```
+
+or into a [JSDoc TypeScript](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html) project with:
+
+```js
+/**
+ * @typedef {import("unist").Node} Node
+ * @typedef {import("unist").Literal} Literal
+ * @typedef {import("unist").Parent} Parent
+ */
+```
+
+### The Languages
+
+`unist` provides the building blocks for having consistent structure, but is often more generic than what we as developers and content authors want to work with.
```

Comment:
Lot of complex words. Maybe there’s a better way to describe that unist provides abstract interfaces